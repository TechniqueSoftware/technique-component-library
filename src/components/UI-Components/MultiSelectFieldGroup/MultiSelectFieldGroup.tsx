// @ts-nocheck
import * as React from 'react';
import clsx from 'clsx';
import Fade from '@material-ui/core/Fade';
import Grid from '@material-ui/core/Grid';
import FormHelperText from '@material-ui/core/FormHelperText';
import withStyles from '@material-ui/styles/withStyles';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { Theme } from '@material-ui/core';
import Radio from '@material-ui/core/Radio';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import Typography from '@material-ui/core/Typography';
import RadioGroup from '@material-ui/core/RadioGroup';
import Box from '@material-ui/core/Box';
import MultiSelectSearchList, {
  MultiSelectGroupValueWrapper,
  MultiSelectOptionGroupData,
  MultiSelectSearchListProps
} from './MultiSelectSearchList/MultiSelectSearchList';

const styles = (theme: Theme) => ({
  helperText: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1)
  },
  helperTextError: {
    color: theme.palette.error.main,
  },
  // Empty placeholder, currently only the parent defines the searchContainer class.
  searchContainer: {}
});

export interface MultiSelectRadioInputProps {
  name: string;
  value: string | 'true' | 'false';
  label: string | React.ReactElement<any, any>;
}

export interface MultiSelectFieldGroupProps {
  onRadioGroupChange?: (event: React.ChangeEvent<{}>, value: string) => void;
  classes?: {
    searchContainer?: string;
    helperText?: string;
    helperTextError?: string;
    error?: string;
  };
  radioGroupProps?: {
    label: string;
    name: string;
    value: string;
  };
  showHelperText?: boolean;
  helperText?: React.ReactNode;
  error?: boolean;
  selectAllRadioInputProps?: MultiSelectRadioInputProps;
  selectSomeRadioInputProps?: MultiSelectRadioInputProps;
  showMultiSelectSearchList?: boolean;
  fieldSetLabelText?: string | React.ReactNode;
  showFieldSetLabelText?: boolean;
  multiSelectOptionGroupDataList: MultiSelectOptionGroupData[];
  groupValueWrapperList: MultiSelectGroupValueWrapper[];
  onCheckboxGroupChange: (groupValues: MultiSelectGroupValueWrapper[]) => void;
  showRadioGroup?: boolean;
  listContainerHeight?: number;
  readOnly?: boolean;
}

export class MultiSelectFieldGroup extends React.Component<MultiSelectFieldGroupProps> {

  static defaultProps: Partial<MultiSelectFieldGroupProps> = {
    showRadioGroup: false,
    showFieldSetLabelText: false,
    showHelperText: true
  };

  constructor(props: MultiSelectFieldGroupProps) {
    super(props);
    this.handleRadioGroupChange = this.handleRadioGroupChange.bind(this);
  }

  renderRadioGroupOption(props: MultiSelectRadioInputProps) {
    const { readOnly } = this.props;
    const { value, label, name } = props;
    return (
      <FormControlLabel
        control={<Radio color={'primary'} />}
        label={label}
        value={value}
        name={name}
        disabled={readOnly} />
    );
  }

  handleRadioGroupChange(event: any, value: string) {
    const {
      onRadioGroupChange,
      onCheckboxGroupChange,
      groupValueWrapperList
    } = this.props;

    const copiedAndEmptiedGroupValues = groupValueWrapperList.map((group) => {
      const copiedGroup = { ...group };
      copiedGroup.values.clear();
      return group;
    });

    onCheckboxGroupChange(copiedAndEmptiedGroupValues);
    onRadioGroupChange(event, value);
  }

  renderRadioGroup() {
    const {
      selectAllRadioInputProps,
      selectSomeRadioInputProps,
      radioGroupProps,
      showRadioGroup,
    } = this.props;

    if (!showRadioGroup) return null;

    return (
      <FormControl component={'fieldset'}>
        <RadioGroup
          aria-label={radioGroupProps.label}
          name={radioGroupProps.name}
          value={radioGroupProps.value}
          onChange={this.handleRadioGroupChange}
        >
          {this.renderRadioGroupOption(selectAllRadioInputProps)}
          {this.renderRadioGroupOption(selectSomeRadioInputProps)}
        </RadioGroup>
      </FormControl>
    );
  }

  render() {
    const {
      classes,
      readOnly,
      error = false,
      multiSelectOptionGroupDataList,
      groupValueWrapperList,
      radioGroupProps,
      selectSomeRadioInputProps,
      showHelperText,
      helperText,
      fieldSetLabelText,
      showFieldSetLabelText,
      listContainerHeight,
      showMultiSelectSearchList,
      onCheckboxGroupChange,
    } = this.props;

    const someRadioOptionSelected = (showMultiSelectSearchList || (radioGroupProps && selectSomeRadioInputProps
       && radioGroupProps.value === selectSomeRadioInputProps.value));

    const multiSelectSearchListProps: MultiSelectSearchListProps = {
      readOnly,
      error,
      groupValueWrapperList,
      multiSelectOptionGroupDataList,
      onCheckboxGroupChange,
      listContainerHeight,
    };

    const fieldSetVariant = showFieldSetLabelText ? undefined : 'srOnly';
    return (
      <React.Fragment>
        <Box component={'fieldset'} border="none" p={0} m={0}>
          <Typography variant={fieldSetVariant} component="span">
            <FormLabel required component={'legend'}>
              {fieldSetLabelText}
            </FormLabel>
          </Typography>
          <Grid item={true} xs={12}>
            {this.renderRadioGroup()}
          </Grid>
          {someRadioOptionSelected &&
          <Fade in={someRadioOptionSelected}>
            <Grid item className={clsx(classes.searchContainer)}>
              {showHelperText &&
              <FormHelperText
                className={clsx(classes.helperText, error && classes.helperTextError)}
              >{helperText}</FormHelperText>
              }
              <MultiSelectSearchList {...multiSelectSearchListProps} />
            </Grid>
          </Fade>
          }
        </Box>
      </React.Fragment>);
  }
}

export default withStyles(styles)(MultiSelectFieldGroup);
