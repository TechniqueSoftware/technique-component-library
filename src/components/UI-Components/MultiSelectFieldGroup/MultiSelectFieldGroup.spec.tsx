import * as React from 'react';
import { shallow } from 'enzyme';
import { MultiSelectFieldGroup, MultiSelectFieldGroupProps } from './MultiSelectFieldGroup';

describe('MultiSelectFieldGroup', () => {
  const baseProps: MultiSelectFieldGroupProps = {
    showHelperText: true,
    showFieldSetLabelText: false,
    fieldSetLabelText: 'mockFieldsetLabelText',
    classes: {
      searchContainer: 'mockSearchContainerClass'
    },
    selectAllRadioInputProps: {
      label: 'mockSelectSomeLabel',
      name: 'mockName',
      value: 'true'
    },
    selectSomeRadioInputProps: {
      label: 'mockSelectSomeLabel',
      name: 'mockName',
      value: 'false'
    },
    radioGroupProps: { name: 'mockRadioGroupName', label: 'mockRadioGroupLabel', value: 'false' },
    onCheckboxGroupChange: jest.fn(),
    onRadioGroupChange: jest.fn(),
    showRadioGroup: false,
    helperText: 'mockHelperText',
    error: false,
    groupValueWrapperList: [{ values: new Set(['1', '2', '3']), groupId: 'group1' }],
    multiSelectOptionGroupDataList: [{
      groupId: 'group1',
      groupLabel: 'Group 1',
      multiSelectOptionDataList: [{
        labelComponent: undefined,
        name: 'mockLabelName',
        value: '1',
        labelString: 'group1-mockLabelText1'
      }, {
        labelComponent: undefined,
        name: 'mockLabelName2',
        value: '2',
        labelString: 'group1-mockLabelText2'
      }
      ]
    },
      {
        groupId: 'group2',
        groupLabel: 'Group 2',
        multiSelectOptionDataList: [{
          labelComponent: undefined,
          name: 'mockLabelName',
          value: '3',
          labelString: 'group2-mockLabelText1'
        }, {
          labelComponent: undefined,
          name: 'mockLabelName2',
          value: '4',
          labelString: 'group2-mockLabelText2'
        }
        ]
      }]
  };

  test('should render with minimal baseProps', () => {
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...baseProps} />);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should show fieldSetLabelText when showFieldSetLabelText is true', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      fieldSetLabelText: 'mockFieldSetLabelText',
      showFieldSetLabelText: true
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);
    expect(shallowWrapper.text()).toContain(props.fieldSetLabelText);
  });

  test('should use srOnly class for fieldSetLabelText when showFieldSetLabelText is false', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      fieldSetLabelText: 'mockFieldSetLabelText',
      showFieldSetLabelText: false
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);
    const srOnlyElement = shallowWrapper.findWhere((node) => node.props().variant === 'srOnly' &&
      node.text() === props.fieldSetLabelText);
    expect(srOnlyElement.length).toBe(1);
  });

  test('should render radioGroup when showRadioGroup is true', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      showRadioGroup: true
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);
    const radioGroup = shallowWrapper.find('ForwardRef(RadioGroup)').parents('WithStyles(ForwardRef(Grid))');
    expect(radioGroup.length).toBe(1);
    expect(radioGroup.shallow().debug()).toMatchSnapshot();
  });

  test('should not render radioGroup when showRadioGroup is false', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      showRadioGroup: false
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);
    const radioGroup = shallowWrapper.find('RadioGroup');
    expect(radioGroup.length).toBe(0);
  });

  test('should show list when the selectSome value is selected', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      showRadioGroup: true,
      radioGroupProps: {
        ...baseProps.radioGroupProps,
        value: baseProps.selectSomeRadioInputProps.value
      }
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);

    const searchContainerShallowWrapper = shallowWrapper.find(`.${baseProps.classes.searchContainer}`);
    const list = searchContainerShallowWrapper.find('WithStyles(MultiSelectSearchList)');
    expect(searchContainerShallowWrapper.debug()).toMatchSnapshot();
    expect(list.length).toBe(1);
  });

  test('should not show list when the selectAll value is selected', () => {
    const props: MultiSelectFieldGroupProps = {
      ...baseProps,
      showRadioGroup: true,
      radioGroupProps: {
        ...baseProps.radioGroupProps,
        value: baseProps.selectAllRadioInputProps.value
      }
    };
    const shallowWrapper = shallow(<MultiSelectFieldGroup {...props} />);

    const searchContainerShallowWrapper = shallowWrapper.find(`.${baseProps.classes.searchContainer}`);
    const list = searchContainerShallowWrapper.find('WithStyles(List)');
    expect(list.length).toBe(0);
    expect(searchContainerShallowWrapper.debug()).toMatchSnapshot();
  });
});
