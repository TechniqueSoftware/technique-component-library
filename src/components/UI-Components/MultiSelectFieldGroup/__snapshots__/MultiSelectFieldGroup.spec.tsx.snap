// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectFieldGroup should not show list when the selectAll value is selected 1`] = `""`;

exports[`MultiSelectFieldGroup should render radioGroup when showRadioGroup is true 1`] = `
"<ForwardRef(Grid) classes={{...}} item={true} xs={12}>
  <WithStyles(ForwardRef(FormControl)) component=\\"fieldset\\">
    <ForwardRef(RadioGroup) aria-label=\\"mockRadioGroupLabel\\" name=\\"mockRadioGroupName\\" value=\\"false\\" onChange={[Function: bound ]}>
      <WithStyles(ForwardRef(FormControlLabel)) control={{...}} label=\\"mockSelectSomeLabel\\" value=\\"true\\" name=\\"mockName\\" disabled={[undefined]} />
      <WithStyles(ForwardRef(FormControlLabel)) control={{...}} label=\\"mockSelectSomeLabel\\" value=\\"false\\" name=\\"mockName\\" disabled={[undefined]} />
    </ForwardRef(RadioGroup)>
  </WithStyles(ForwardRef(FormControl))>
</ForwardRef(Grid)>"
`;

exports[`MultiSelectFieldGroup should render with minimal baseProps 1`] = `
"<Fragment>
  <Styled(MuiBox) component=\\"fieldset\\" border=\\"none\\" p={0} m={0}>
    <WithStyles(ForwardRef(Typography)) variant=\\"srOnly\\" component=\\"span\\">
      <WithStyles(ForwardRef(FormLabel)) required={true} component=\\"legend\\">
        mockFieldsetLabelText
      </WithStyles(ForwardRef(FormLabel))>
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Grid)) item={true} xs={12} />
    <ForwardRef(Fade) in={true}>
      <WithStyles(ForwardRef(Grid)) item={true} className=\\"mockSearchContainerClass\\">
        <WithStyles(ForwardRef(FormHelperText)) className=\\"\\">
          mockHelperText
        </WithStyles(ForwardRef(FormHelperText))>
        <WithStyles(MultiSelectSearchList) readOnly={[undefined]} error={false} groupValueWrapperList={{...}} multiSelectOptionGroupDataList={{...}} onCheckboxGroupChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} listContainerHeight={[undefined]} />
      </WithStyles(ForwardRef(Grid))>
    </ForwardRef(Fade)>
  </Styled(MuiBox)>
</Fragment>"
`;

exports[`MultiSelectFieldGroup should show list when the selectSome value is selected 1`] = `
"<WithStyles(ForwardRef(Grid)) item={true} className=\\"mockSearchContainerClass\\">
  <WithStyles(ForwardRef(FormHelperText)) className=\\"\\">
    mockHelperText
  </WithStyles(ForwardRef(FormHelperText))>
  <WithStyles(MultiSelectSearchList) readOnly={[undefined]} error={false} groupValueWrapperList={{...}} multiSelectOptionGroupDataList={{...}} onCheckboxGroupChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} listContainerHeight={[undefined]} />
</WithStyles(ForwardRef(Grid))>"
`;
